Mathlib/SetNotation.lean

```diff
@@ -0,0 +1,40 @@
+class Mem (α : outParam $ Type u) (γ : Type v) where
+  mem : α → γ → Prop
+
+infix:50 " ∈ " => Mem.mem
+
+notation:50 x " ∉ " s => ¬ x ∈ s
+
+class Subset (α : Type u) where
+  subset : α → α → Prop
+
+infix:50 " ⊆ " => Subset.subset
+
+class Union (α : Type u) where
+  union : α → α → α
+
+infixl:65 " ∪ " => Union.union
+
+class Inter (α : Type u) where
+  inter : α → α → α
+
+infixl:70 " ∩ " => Inter.inter
+
+class Sdiff (α : Type u) where
+  sdiff : α → α → α
+
+infix:70 " \\ " => Sdiff.sdiff
+
+declare_syntax_cat binderterm -- notation for `a` or `a : A` or `a ∈ S`
+syntax ident : binderterm
+syntax ident " : " term : binderterm
```

Comment:
```suggestion
```

If you add syntax for `∀ x : α, ...`, then you should add a macro/elaborator for it as well.  The only reason this works at the moment is pure coincidence: the built-in parser for forall is built in the same way, and therefore the built-in expander for forall also happens to apply to our newly defined syntax.

This overlap also produces really ugly error messages:
```lean
#check ∀ x : Foo, x = x
/-
▶ 42:8-42:24: error:
overloaded, errors 
  elaboration function for '«term∀_,_»' has not been implemented
    ∀x : Foo,x = x
  
  42:13 unknown identifier 'Foo'
-/
```

Reply:
I only moved this declaration from `Set.lean`. It looks like @kbuzzard put it there. Why do we need it, if the built-in expander for forall already handles it? Should I simply delete it and add syntax only for the bounded versions?

Also: I experimented with a delaborator for bounded quantifiers. It seems that, at least for forall, we need to introduce a definition like `forallb` to register it. Should I do that, or just let the pretty-printer print the expanded version?

Reply:
Yes, I think the `ident ":" term : binderterm` syntax should just be removed (at least for now).

> I experimented with a delaborator for bounded quantifiers. It seems that, at least for forall, we need to introduce a definition like forallb to register it.

We might want `forallb` for other reasons (e.g. to ensure that simp doesn't simplify a bounded quantifier `∀ x < max a b, p x` to a non-bounded quantifier like `∀ x, x < a ∨ x < b → p x`).  But it's easy enough to register a delaborator for forall expressions:
```lean
import Lean

open Lean PrettyPrinter.Delaborator

@[delab forallE]
def delabForallBounded : Delab :=
  -- use `guard` here to restrict to bounded quantifiers
  `(here goes your beautiful output)

#check ∀ x, true -- here✝ goes✝ your✝ beautiful✝ output✝ : Prop
```

Reply:
Thanks, it's good to know that we can match against `forallE`. I didn't add the delaborator, but I just made all the other corrections.

Reply:
I might have PR'ed that binder stuff but I strongly suspect some kids on the discord wrote it, and I don't claim that either I or they knew what they were doing :-) The file `Set.lean` was stuff ported from core Lean 3 which was not in core Lean 4 (so I didn't really know where to put it in this repo). I did the same thing with Function.lean . I didn't know if we were adopting a "please keep all Lean 4 mathlib files as a proper subset of the corresponding Lean 3 mathlib files" attitude or not.